src/OrleansRuntime/Core/InsideRuntimeClient.cs

```diff
@@ -364,35 +364,50 @@ public void SniffIncomingMessage(Message message)
                         throw exc;
                     }
 
-                    // If the target has a grain-level interceptor or there is a silo-level interceptor, intercept the call.
-                    var shouldCallSiloWideInterceptor = SiloProviderRuntime.Instance.GetInvokeInterceptor() != null && target is IGrain;
+                    Task<object> resultTask;
+
+                    // If the target has a grain-level interceptor or there is a silo-level interceptor, intercept the
+                    // call.
+                    var shouldCallSiloWideInterceptor = SiloProviderRuntime.Instance.GetInvokeInterceptor() != null
+                                                        && target is IGrain;
                     var intercepted = target as IGrainInvokeInterceptor;
-                    if (intercepted != null || shouldCallSiloWideInterceptor)
+                    var grainHasInterceptor = intercepted != null;
+                    if (grainHasInterceptor || shouldCallSiloWideInterceptor)
                     {
-                        // Fetch the method info for the intercepted call.
-                        var implementationInvoker =
-                            invocationMethodInfoMap.GetInterceptedMethodInvoker(target.GetType(), request.InterfaceId,
+                        // Get an invoker which delegates to the grain's IGrainInvocationInterceptor implementation.
+                        // If the grain does not implement IGrainInvocationInterceptor, then the invoker simply
+                        // delegates calls to the provided invoker.
+                        var interceptedInvoker =
+                            this.interceptedMethodInvokerCache.GetOrCreate(
+                                target.GetType(),
+                                request.InterfaceId,
                                 invoker);
-                        var methodInfo = implementationInvoker.GetMethodInfo(request.MethodId);
+                        var methodInfo = interceptedInvoker.GetMethodInfo(request.MethodId);
                         if (shouldCallSiloWideInterceptor)
                         {
                             // There is a silo-level interceptor and possibly a grain-level interceptor.
-                            var runtime = SiloProviderRuntime.Instance;
-                            resultObject =
-                                await runtime.CallInvokeInterceptor(methodInfo, request, target, implementationInvoker);
+                            // As a minor optimization, only pass the intercepted invoker if there is a grain-level
+                            // interceptor.
+                            resultTask = SiloProviderRuntime.Instance.CallInvokeInterceptor(
+                                methodInfo,
+                                request,
+                                target,
+                                grainHasInterceptor ? interceptedInvoker : invoker);
```

Comment:
I don t understand this logic. why we sometimes call interceptedInvoker and sometimes invoker.
is it the case that you call both global and then per grain interceptor and then the actual call? 3 step call? Also, confused with the global if and then sub case if else.